cdap-app-fabric/src/main/java/co/cask/cdap/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,177 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.runtime.monitor;
+
+import co.cask.cdap.api.messaging.MessagePublisher;
+import co.cask.cdap.client.config.ClientConfig;
+import co.cask.cdap.client.util.RESTClient;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.logging.LogSampler;
+import co.cask.cdap.common.logging.LogSamplers;
+import co.cask.cdap.common.logging.Loggers;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.common.http.HttpMethod;
+import co.cask.common.http.HttpResponse;
+import com.google.common.base.Supplier;
+import com.google.common.reflect.TypeToken;
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.ServiceUnavailableException;
+
+/**
+ * Runtime Monitor Runnable responsible for fetching metadata
+ */
+public class RuntimeMonitor extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeMonitor.class);
+  // For outage, only log once per 60 seconds per message.
+  private static final Logger OUTAGE_LOG = Loggers.sampling(LOG, LogSamplers.perMessage(new Supplier<LogSampler>() {
+    @Override
+    public LogSampler get() {
+      return LogSamplers.limitRate(60000);
+    }
+  }));
+
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final Type MAP_STRING_MESSAGE_TYPE = new TypeToken<Map<String, List<MonitorMessage>>>() { }.getType();
+
+  private final RESTClient restClient;
+  private final ClientConfig clientConfig;
+
+  private final int limit;
+  private final String runtimeId;
+  private final CConfiguration cConf;
+  private Map<String, MonitorConsumeRequest> topicsToRequest;
+
+  private final MessagePublisher messagePublisher;
+  private final long pollTimeMillis;
+  private volatile Thread runThread;
+  private volatile boolean stopped;
+
+  public RuntimeMonitor(String runtimeId, CConfiguration cConf, MessagePublisher messagePublisher,
+                        ClientConfig clientConfig) {
+    this.runtimeId = runtimeId;
```

Comment:
What is runtimeId??

Reply:
This is the id for each runtime, its will be used to distinguish runtimes while storing offsets.

Reply:
Shouldn't it be `ProgramRunId`?